include/seqan3/range/view/kmer_hash.hpp

```diff
@@ -12,47 +12,645 @@
+     *
+     * \details
+     *
+     * This element acts as a placeholder; attempting to dereference it results in undefined behaviour.
+     *
+     * ### Complexity
+     *
+     * Constant.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    auto end() noexcept
+    {
+        return std::ranges::end(urange);
+    }
+
+    //!\copydoc end()
+    auto end() const noexcept
+    //!\cond
+        requires ConstIterableRange<urng_t>
+    //!\endcond
+    {
+        return std::ranges::end(urange);
+    }
+
+    //!\copydoc end()
+    auto cend() const noexcept
+    //!\cond
+        requires ConstIterableRange<urng_t>
+    //!\endcond
+    {
+        return std::ranges::end(urange);
+    }
+    //!\}
+};
+
+//!\brief A deduction guide for the view class template.
+template <std::ranges::ViewableRange orng_t>
+kmer_hash_view(orng_t &&, shape const & shape_) -> kmer_hash_view<std::ranges::all_view<orng_t>>;
+
+// ---------------------------------------------------------------------------------------------------------------------
+// kmer_hash_fn (adaptor definition)
+// ---------------------------------------------------------------------------------------------------------------------
+
+//![adaptor_def]
+//!\brief view::kmer_hash's range adaptor object type (non-closure).
+struct kmer_hash_fn
+{
+    /*!\brief            Call the view's constructor with the underlying view and a seqan3::shape as argument.
      * \param[in] urange The input range to process. Must model std::ranges::ViewableRange and the reference type of the
      *                   range of the range must model seqan3::Semialphabet.
-     * \param[in] k      The k-mer size to construct hashes for.
+     * \param[in] shape_ The seqan3::shape to use for hashing.
+     * \throws std::invalid_argument if resulting hash values would be too big for a 64 bit integer.
      * \returns          A range of converted elements.
      */
-    template <std::ranges::ViewableRange urng_t>
-    //!\cond
-        requires Semialphabet<reference_t<urng_t>>
-    //!\endcond
-    constexpr auto operator()(urng_t && urange, size_t const k) const noexcept
+    template <std::ranges::Range urng_t>
+    constexpr auto operator()(urng_t && urange, shape const & shape_) const
     {
-        return std::forward<urng_t>(urange) | ranges::view::sliding(k) | std::view::transform(
-        [] (auto const in)
+        static_assert(std::ranges::ViewableRange<urng_t>,
+            "The range parameter to view::kmer_hash cannot be a temporary of a non-view range.");
+        static_assert(std::ranges::ForwardRange<urng_t>,
+            "The range parameter to view::kmer_hash must model std::ranges::ForwardRange.");
+        static_assert(Semialphabet<reference_t<urng_t>>,
+            "The range parameter to view::kmer_hash must be over elements of seqan3::Semialphabet.");
+
+        if (std::ceil(std::log2(alphabet_size<reference_t<urng_t>>)) * shape_.size() > 64u)
```

Comment:
Is 64 the right number?

Reply:
Yes, the hashes are calculated via arithmetic coding. So I need to ensure that the maximum hash value fits into a `uint64_t`.

**But** actually the correct way is to do `shape_.size() <= (64 / std::log2(alphabet_size<reference_t<urng_t>>))`.